Mathlib/Data/Real/OfDigits.lean

```diff
@@ -0,0 +1,218 @@
+  simp only [ofDigitsTerm]
+  positivity
+
+private lemma b_pos {b : ℕ} (digits : ℕ → Fin b) : 0 < b := by
+  rw [pos_iff_ne_zero]
+  rintro rfl
+  exact IsEmpty.false digits
+
+theorem ofDigitsTerm_le {b : ℕ} {digits : ℕ → Fin b} {n : ℕ} :
+    ofDigitsTerm digits n ≤ (b - 1) * ((b : ℝ) ^ (n + 1))⁻¹ := by
+  have hb : 0 < b := b_pos digits
+  simp only [ofDigitsTerm]
+  gcongr
+  rw [← Nat.cast_pred hb]
+  norm_cast
+  omega
+
+theorem summable_ofDigitsTerm {b : ℕ} {digits : ℕ → Fin b} :
+    Summable (ofDigitsTerm digits) := by
+  have hb : 0 < b := b_pos digits
+  obtain rfl | hb := (Nat.one_le_of_lt hb).eq_or_lt
+  · have : ofDigitsTerm digits = 0 := by
+      ext i
+      simp [ofDigitsTerm]
+    simp only [this]
+    eta_expand
+    simp [summable_const_iff]
+  have h := summable_geometric_of_lt_one (r := (b⁻¹ : ℝ)) (by simp)
+    (by rify at hb; exact inv_lt_one_of_one_lt₀ hb)
+  apply Summable.mul_left (a := (b : ℝ)) at h
+  replace h : Summable fun i ↦ b * (b : ℝ)⁻¹ ^ (i + 1) := by
+    simp_rw [pow_succ', ← mul_assoc, mul_comm (b : ℝ), mul_assoc]
+    exact Summable.mul_left _ h
+  apply Summable.of_nonneg_of_le _ _ h
+  · intros
+    exact ofDigitsTerm_nonneg
+  intro i
+  apply le_trans ofDigitsTerm_le
+  gcongr
+  · simp
+  · rw [inv_pow]
+
+/-- The number `0.d₀d₁d₂...` in the system with base `b`.
+We allow repeating representations like `0.999...` here. -/
+noncomputable def ofDigits {b : ℕ} (digits : ℕ → Fin b) : ℝ :=
+  ∑' n, ofDigitsTerm digits n
+
+theorem ofDigits_nonneg {b : ℕ} (digits : ℕ → Fin b) : 0 ≤ ofDigits digits := by
+  simp only [ofDigits]
+  exact tsum_nonneg fun _ ↦ ofDigitsTerm_nonneg
+
+theorem ofDigits_le_one {b : ℕ} (digits : ℕ → Fin b) :
+    ofDigits digits ≤ 1 := by
+  have hb : 0 < b := b_pos digits
+  obtain rfl | hb' := (Nat.one_le_of_lt hb).eq_or_lt
+  · simp [ofDigits, ofDigitsTerm]
+  rify at hb'
+  have hb_inv_nonneg : 0 ≤ (b⁻¹ : ℝ) := by simp
+  have hb_inv_lt_one : (b⁻¹ : ℝ) < 1 := inv_lt_one_of_one_lt₀ hb'
+  simp only [ofDigits]
+  let g (i : ℕ) : ℝ := (1 - (b⁻¹ : ℝ)) * (b⁻¹ : ℝ) ^ i
+  have hg_summable : Summable g :=
+    (summable_geometric_of_lt_one (by simp) (inv_lt_one_of_one_lt₀ hb')).mul_left _
+  convert Summable.tsum_mono (summable_ofDigitsTerm) hg_summable _
+  · simp [g, tsum_mul_left, -inv_pow]
+    rw [tsum_geometric_of_lt_one hb_inv_nonneg hb_inv_lt_one, mul_inv_cancel₀]
+    linarith
+  · intro i
+    simp only [inv_pow, g]
+    convert ofDigitsTerm_le using 1
+    field_simp
+    left
+    rw [pow_succ']
+
+theorem ofDigits_eq_partial_sum_add_ofDigits {b : ℕ} (a : ℕ → Fin b) (n : ℕ) :
+    ofDigits a = (∑ i ∈ Finset.range n, ofDigitsTerm a i) +
+      ((b : ℝ) ^ n)⁻¹ * ofDigits (fun i ↦ a (i + n)) := by
+  simp only [ofDigits]
+  rw [← Summable.sum_add_tsum_nat_add n summable_ofDigitsTerm,
+    ← Summable.tsum_mul_left _ summable_ofDigitsTerm]
+  congr
+  ext i
+  simp only [ofDigitsTerm]
+  ring
+
+theorem ofDigits_close_of_common_prefix {b : ℕ} {x y : ℕ → Fin b} {n : ℕ}
```

Comment:
`close` is not part of the naming convention either, and neither is `common_prefix`. How about `abs_ofDigits_sub_ofDigits_le`?